### PR TITLE
Segqueue

### PR DESCRIPTION
**Pull request by yoshuawuyts**
Followed [`dignifiedquire/byte-pool`](https://github.com/dignifiedquire/byte-pool)'s lead and used crossbeam's `SegQueue` instead, replacing the Treiber stack. This means we no longer have unsafe code, but more importantly: it improves perf by about 2-3x on our benches. Thanks!

## Before
```txt
test remem::contention    ... bench:   1,564,555 ns/iter (+/- 205,107)
test remem::create        ... bench:          47 ns/iter (+/- 0)
test remem::no_contention ... bench:     309,059 ns/iter (+/- 155,469)
```

## After
```txt
test remem::contention        ... bench:     915,251 ns/iter (+/- 17,242)
test remem::create            ... bench:         115 ns/iter (+/- 0)
test remem::no_contention     ... bench:     100,734 ns/iter (+/- 2,801)
```

### CODE REVIEW

File: benches/pool.rs [hunk 1]
```diff
@@ -26,14 +26,58 @@ mod remem {
...
+mod byte_pool {
+    use super::CAPACITY;
+    use byte_pool::BytePool;
+    use std::sync::Arc;
+    use std::thread;
+    use test::{black_box, Bencher};
+
+    #[bench]
+    fn create(b: &mut Bencher) {
+        b.iter(|| Arc::new(BytePool::new()));
```
Comment (dignifiedquire) [inline benches/pool.rs:59]: this is measuring pool creation, not actual object allocations

Comment (yoshuawuyts) [reply benches/pool.rs:59]: oh yeah, that's what it's intended to measure haha. (which is to say: this is the intended behavior).

Comment (yoshuawuyts) [reply benches/pool.rs:59]: Following async-std's lead here: https://github.com/async-rs/async-std/blob/master/benches/mutex.rs#L9-L12

Comment (dignifiedquire) [reply benches/pool.rs:59]: In that case I would argue this should be set to `b.iter(|| {})` for the `Vec` case, as it has no overhead, as there is no pool being created

Comment (rylev) [reply benches/pool.rs:59]: Yea I'm also not sure that the create benches all measure the same thing. Some are fully initialized and some are not

**Review by rylev — APPROVED** [review]
Small nit about benchmarks but that can be addressed in another PR. Looks great! Thanks ❤️

**Pull request merged**